File: packages/protocol/src/schedule.ts
const MAX_UINT64 = (1n << 64n) - 1n;
const MAX_UINT32 = (1n << 32n) - 1n;

export interface ScheduleWindowV1 {
  slot: bigint;
  deadline: bigint;
}

/** Computes an inclusive UTC slot window using checked wire-width arithmetic. */
export function scheduleWindowV1(
  startAt: bigint,
  intervalSeconds: bigint,
  graceSeconds: bigint,
  occurrence: bigint,
): ScheduleWindowV1 | null {
  if (startAt < 0n || startAt > MAX_UINT64
    || intervalSeconds <= 0n || intervalSeconds > MAX_UINT64
    || graceSeconds <= 0n || graceSeconds >= intervalSeconds || graceSeconds > MAX_UINT64
    || occurrence <= 0n || occurrence > MAX_UINT32) return null;
  const index = occurrence - 1n;
  if (index !== 0n && intervalSeconds > (MAX_UINT64 - startAt) / index) return null;
  const slot = startAt + index * intervalSeconds;
  if (graceSeconds > MAX_UINT64 - slot) return null;
  return { slot, deadline: slot + graceSeconds };
}

File: apps/fcc-extension/internal/protocol/schedule.go
package protocol

import "math"

type ScheduleWindowResultV1 struct {
	Slot     uint64
	Deadline uint64
}

// ScheduleWindowV1 computes an inclusive UTC slot window without overflowing
// the uint64 wire timestamp domain.
func ScheduleWindowV1(startAt, intervalSeconds, graceSeconds, occurrence uint64) (ScheduleWindowResultV1, bool) {
	if intervalSeconds == 0 || graceSeconds == 0 || graceSeconds >= intervalSeconds || occurrence == 0 || occurrence > math.MaxUint32 {
		return ScheduleWindowResultV1{}, false
	}
	index := occurrence - 1
	if index != 0 && intervalSeconds > (math.MaxUint64-startAt)/index {
		return ScheduleWindowResultV1{}, false
	}
	slot := startAt + index*intervalSeconds
	if graceSeconds > math.MaxUint64-slot {
		return ScheduleWindowResultV1{}, false
	}
	return ScheduleWindowResultV1{Slot: slot, Deadline: slot + graceSeconds}, true
}
